src/lib/optimizer/strategy/join_to_predicate_rewrite_rule.cpp

```diff
@@ -0,0 +1,187 @@
+  // If we find an equals-predicate that filters on a UCC, a maximum of one line is singled out in the result relation.
+  // Since at this point, we already know the join in question is basically a semi-join, we can further transform the
+  // join to a single predicate node filtering out this one line testing against values in the join column.
+  visit_lqp(removable_subtree, [&removable_subtree, &valid_predicate](auto& current_node) {
+    if (current_node->type == LQPNodeType::Union) {
+      return LQPVisitation::DoNotVisitInputs;
+    }
+    if (current_node->type != LQPNodeType::Predicate) {
+      return LQPVisitation::VisitInputs;
+    }
+
+    const auto candidate = std::static_pointer_cast<PredicateNode>(current_node);
+    const auto candidate_exp = std::dynamic_pointer_cast<BinaryPredicateExpression>(candidate->predicate());
+
+    if (!candidate_exp) {
+      return LQPVisitation::VisitInputs;
+    }
+
+    Assert(candidate_exp, "Need to get a predicate with a BinaryPredicateExpression.");
+
+    // Only predicates that filter using equals cond. on a column with a constant value are of use to our optimization.
+    // These conditions have the potential (given filtered column is a UCC) to single out a maximum of one result tuple.
+    if (candidate_exp->predicate_condition != PredicateCondition::Equals)
+      return LQPVisitation::VisitInputs;
+
+    std::shared_ptr<AbstractExpression> candidate_column_expr = nullptr;
+    std::shared_ptr<AbstractExpression> candidate_value_expr = nullptr;
+
+    for (const auto& predicate_operand : {candidate_exp->left_operand(), candidate_exp->right_operand()}) {
+      if (predicate_operand->type == ExpressionType::LQPColumn) {
+        candidate_column_expr = predicate_operand;
+      } else if (predicate_operand->type == ExpressionType::Value) {
+        candidate_value_expr = predicate_operand;
+      }
+    }
+
+    // There should not be a case where we run into no column, but there may be a case where we have no value but
+    // compare columns.
+    if (!candidate_column_expr || !candidate_value_expr) {
+      return LQPVisitation::VisitInputs;
+    }
+
+    // If the Predicate column expr != join column expression, check whether the column referenced is still available
+    // in the join.
+    if (!expression_evaluable_on_lqp(candidate_column_expr, *removable_subtree)) {
+      return LQPVisitation::VisitInputs;
+    }
+
+    // Check for uniqueness.
+    const auto testable_expressions = ExpressionUnorderedSet{candidate_column_expr};
+
+    if (!removable_subtree->has_matching_unique_constraint(testable_expressions)) {
+      return LQPVisitation::VisitInputs;
+    }
+
+    valid_predicate = candidate;
+    return LQPVisitation::DoNotVisitInputs;
+  });
+
+  if (valid_predicate) {
+    auto entry = std::make_tuple(join_node, removable_side.value(), valid_predicate);
+    rewritables.emplace_back(entry);
```

Comment:
Does `rewritables.emplace_back(join_node, removable_side.value(), valid_predicate);` not work?